docs/process/data_versioning.md

```diff
@@ -0,0 +1,103 @@
+
+The following strategy should be employed:
+
+1. **Old Code, New Data:** Datagen can add obsolete resource keys to the data file in order to support a particular minimum version of ICU4X. Depending on the datagen configuration, it should be possible to support ***both the current and the previous major version***. For example, a data file built for ICU4X 3.7.1 may choose to support versions down to ICU4X 2.0.0. This may result in increased data size, since both old data and new data are shipped in parallel; however, shared data files are designed to reduce data size overall.
+2. **New Code, Old Data:** Stable ICU4X code should be able to read from data files built for any ICU4X version ***with the same major version number***. The goal is always to provide a "best effort" result. New code has several options when consuming old data:
+    1. Map the old data struct to the new data struct at runtime
+    2. If (1) is infeasible, retain code required to evaluate the old struct
+    3. Use the Default impl for the new data struct
+
+In addition, new data structs should only be added in minor releases, not patch releases.
+
+#### Data File Visualization
+
+We can visualize this model as follows:
+
+| Code Versions Supported | 1.0 | 1.1 | 1.2 | 2.0 | 2.1 | 2.2 |
+|---|---|---|---|---|---|---|
+| Data @ 1.0 | ✔️ | ☑️ | ☑️ | ❌ | ❌ | ❌ |
+| Data @ 1.2 | ❌ | ❌ | ✔️ | ❌ | ❌ | ❌ |
+| Data @ 1.2 w/ 1.0 compat | ✔️ | ✔️ | ✔️ | ❌ | ❌ | ❌ |
+| Data @ 2.0 | ❌ | ❌ | ❌ | ✔️ | ☑️ | ☑️ |
+| Data @ 2.0 w/ 1.0 compat | ✔️ | ✔️ | ✔️ | ✔️ | ☑️ | ☑️ |
+| Data @ 2.2 | ❌ | ❌ | ❌ | ❌ | ❌ | ✔️ |
+| Data @ 2.2 w/ 2.0 compat | ❌ | ❌ | ❌ | ✔️ | ✔️ | ✔️ |
+| Data @ 2.2 w/ 1.0 compat | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ | ✔️ |
+
+- ✔️ = supported in all constructors
+- ☑️ = supported in compatibility constructors
+- ❌ = not supported
+
+#### Blob Version versus Data Struct Version
+
+Adding older data structs to a newer data file is easy. How do we handle cases where we want to change the top-level blob layout? There are two cases here:
+
+1. New binary format (e.g. Postcard 2.0)
+2. New blob layout (e.g. change ZeroMap to ZeroHashMap)
+
+In case 1, the new binary format should be added as a new syntax feature, parallel to the `postcard_1`, `bincode_1`, and `json` features.
+
+In case 2, the `BlobSchema` will need to gain a new variant. When building data files with compatibility for older code versions, it will need to use a `BlobSchema` variant that is supported by all code versions that are intended to be supported.
+
+### IV. Constructor Versioning
+
+*Also see: [icu_provider::constructors](https://icu4x.unicode.org/doc/icu_provider/constructors/index.html)*
+
+All ICU4X functions that take a data provider should expose three signatures:
+
+- `*_unstable` works only with an _exactly matching_ data version.
+- `*_with_any_provider` and `*_with_buffer_provider` are _compatibility constructors_ supporting the current data version and all data versions from the current major release.
```

Comment:
suggestion: also mention that we may add additional compat constructors for prior releases?

Reply:
Added a note